File: windows/CursorTokenCore/AccountValidity.cs
using System.Globalization;

namespace CursorTokenCore;

public static class AccountValidity
{
    public const string LongTerm = "long_term";
    public const string Temporary = "temporary";
    public const int MaxDays = 999;
    public const int MaxHours = 23;

    public static string SanitizeKind(string? raw)
    {
        var key = (raw ?? "").Trim().ToLowerInvariant().Replace('-', '_');
        return key is "temporary" or "temp" or "short" ? Temporary : LongTerm;
    }

    public static int ClampDays(object? raw) => ClampInt(raw, 0, MaxDays);
    public static int ClampHours(object? raw) => ClampInt(raw, 0, MaxHours);

    public static bool IsTemporary(Account? account) =>
        account is not null && SanitizeKind(account.AccountKind) == Temporary;

    public static string? ComputeEndIso(string? startAt, int days, int hours)
    {
        var start = AccountSync.ParseIso(startAt);
        if (start is null) return null;
        var d = ClampDays(days);
        var h = ClampHours(hours);
        if (d == 0 && h == 0) return null;
        return AccountSync.NowIso(start.Value.AddDays(d).AddHours(h));
    }

    public static string? AccountEndIso(Account? account)
    {
        if (!IsTemporary(account) || account is null) return null;
        return ComputeEndIso(account.TempStartAt, account.TempValidDays, account.TempValidHours);
    }

    public static void ApplyEndOverride(UsageSnapshot snap, Account? account, DateTimeOffset? now = null)
    {
        var end = AccountEndIso(account);
        if (string.IsNullOrEmpty(end)) return;
        snap.BillingCycleEnd = end;
        snap.DaysRemaining = UsageParser.DaysUntil(end, now ?? DateTimeOffset.UtcNow);
        snap.BillingCycleEndOverridden = true;
    }

    public static bool ValidityEquals(Account a, string kind, string startAt, int days, int hours) =>
        SanitizeKind(a.AccountKind) == SanitizeKind(kind)
        && (a.TempStartAt ?? "").Trim() == (startAt ?? "").Trim()
        && ClampDays(a.TempValidDays) == ClampDays(days)
        && ClampHours(a.TempValidHours) == ClampHours(hours);

    static int ClampInt(object? raw, int lo, int hi)
    {
        int n;
        switch (raw)
        {
            case int i: n = i; break;
            case long l: n = (int)l; break;
            case double d: n = (int)d; break;
            case string s when int.TryParse(s, NumberStyles.Integer, CultureInfo.InvariantCulture, out var p): n = p; break;
            default: return 0;
        }
        return Math.Clamp(n, lo, hi);
    }
}

File: macos/Sources/CursorTokenCore/AccountValidity.swift
import Foundation

public enum AccountValidity {
    public static let longTerm = "long_term"
    public static let temporary = "temporary"
    public static let maxDays = 999
    public static let maxHours = 23

    public static func sanitizeKind(_ raw: String?) -> String {
        let key = (raw ?? "").trimmingCharacters(in: .whitespaces).lowercased().replacingOccurrences(of: "-", with: "_")
        if key == "temporary" || key == "temp" || key == "short" { return temporary }
        return longTerm
    }

    public static func clampDays(_ raw: Any?) -> Int { clampInt(raw, lo: 0, hi: maxDays) }
    public static func clampHours(_ raw: Any?) -> Int { clampInt(raw, lo: 0, hi: maxHours) }

    public static func isTemporary(_ account: Account?) -> Bool {
        guard let account else { return false }
        return sanitizeKind(account.accountKind) == temporary
    }

    public static func computeEndIso(startAt: String?, days: Int, hours: Int) -> String? {
        guard let start = AccountSync.parseIso(startAt) else { return nil }
        let d = clampDays(days)
        let h = clampHours(hours)
        if d == 0 && h == 0 { return nil }
        let end = start.addingTimeInterval(TimeInterval(d * 86_400 + h * 3_600))
        return AccountSync.nowIso(end)
    }

    public static func accountEndIso(_ account: Account?) -> String? {
        guard isTemporary(account), let account else { return nil }
        return computeEndIso(startAt: account.tempStartAt, days: account.tempValidDays, hours: account.tempValidHours)
    }

    public static func applyEndOverride(_ snap: inout UsageSnapshot, account: Account?, now: Date = Date()) {
        guard let end = accountEndIso(account) else { return }
        snap.billingCycleEnd = end
        snap.daysRemaining = UsageParser.daysUntil(end, now: now)
        snap.billingCycleEndOverridden = true
    }

    static func clampInt(_ raw: Any?, lo: Int, hi: Int) -> Int {
        let n: Int
        if let i = raw as? Int { n = i }
        else if let d = raw as? Double { n = Int(d) }
        else if let s = raw as? String, let v = Int(s) { n = v }
        else if let n0 = raw as? NSNumber { n = n0.intValue }
        else { return 0 }
        return min(hi, max(lo, n))
    }
}
